third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/PingPong.cpp

```diff
@@ -73,202 +274,655 @@ void getNestedFor(Region *partition,
+  CriticalOpType opType = opTypeOpt.value();
+  if (opType == CriticalOpType::NonReorderable) {
+    LDBG("Operation " << opName << " is not reorderable.");
+    // Find the first op with memory side effect after this critical op
+    Operation *curOp = keyOp;
+    while (curOp) {
+      if (!isMemoryEffectFree(curOp)) {
+        LDBG("Found op with memory effects:");
+        dumpMemoryEffects(curOp);
+        return curOp;
+      }
+      // Set end op to the end of the control flow equivalent region
+      if (!areControlFlowEquivalent(curOp, curOp->getNextNode(), domInfo,
+                                    postDomInfo))
+        return curOp;
+      curOp = curOp->getNextNode();
+    }
+  } else if (opType == CriticalOpType::PureArithmetic) {
+    LDBG("Operation " << opName << " is pure arithmetic.");
+    // Find the first op with memory side effect after this critical op
+    Operation *curOp = keyOp;
+    while (curOp) {
+      if (!isMemoryEffectFree(curOp)) {
+        LDBG("Found op with memory effects:");
+        dumpMemoryEffects(curOp);
+        return curOp;
+      }
+      // Set end op to the end of the control flow equivalent region
+      if (!areControlFlowEquivalent(curOp, curOp->getNextNode(), domInfo,
+                                    postDomInfo))
+        return curOp;
+      curOp = curOp->getNextNode();
+    }
+  }
+  return nullptr;
+}
+
+/// Returns the operation from startOps that is closest to the entry
+/// (executed earliest). All ops must be in the same block.
+Operation *unionOfStartOps(SmallVector<Operation *> &startOps) {
+  if (startOps.empty())
+    return nullptr;
+
+  Operation *earliestOp = startOps[0];
+  Block *block = earliestOp->getBlock();
+
+  for (size_t i = 1; i < startOps.size(); ++i) {
+    Operation *op = startOps[i];
+    // Verify all ops are in the same block
+    if (op->getBlock() != block) {
+      LDBG("Warning: unionOfStartOps called with ops in different blocks");
+      return nullptr;
+    }
+    // If op is before earliestOp, then op is closer to entry
+    if (op->isBeforeInBlock(earliestOp)) {
+      earliestOp = op;
+    }
+  }
+  return earliestOp;
+}
+
+/// Returns the operation from endOps that is closest to the terminator
+/// (executed latest). All ops must be in the same block.
+Operation *unionOfEndOps(SmallVector<Operation *> &endOps) {
+  if (endOps.empty())
+    return nullptr;
+
+  Operation *latestOp = endOps[0];
+  Block *block = latestOp->getBlock();
+
+  for (size_t i = 1; i < endOps.size(); ++i) {
+    Operation *op = endOps[i];
+    // Verify all ops are in the same block
+    if (op->getBlock() != block) {
+      LDBG("Warning: unionOfEndOps called with ops in different blocks");
+      return nullptr;
+    }
+    // If latestOp is before op, then op is closer to terminator
+    if (latestOp->isBeforeInBlock(op)) {
+      latestOp = op;
+    }
+  }
+  return latestOp;
+}
+
+static void handleWarpSpec(ttg::WarpSpecializeOp wsOp, int computeCapability) {
```

Comment:
Doc-comment would be helpful, since the name `handleWarpSpec` is not very descriptive.